source/server/admin/stats_handler.cc

```diff
@@ -225,29 +254,82 @@ StatsHandler::statsAsJson(const std::map<std::string, uint64_t>& all_stats,
     stats_array.push_back(ValueUtil::structValue(stat_obj));
   }
 
-  ProtobufWkt::Struct histograms_obj;
-  auto* histograms_obj_fields = histograms_obj.mutable_fields();
-
   ProtobufWkt::Struct histograms_obj_container;
   auto* histograms_obj_container_fields = histograms_obj_container.mutable_fields();
-  std::vector<ProtobufWkt::Value> computed_quantile_array;
 
-  bool found_used_histogram = false;
+  // Display bucket data if histogram_buckets query parameter is used, otherwise output contains
+  // quantile summary data.
+  switch (histogram_buckets_value) {
+  case Utility::HistogramBucketsValue::Null:
+    statsAsJsonQuantileSummaryHelper(*histograms_obj_container_fields, all_histograms, used_only,
+                                     regex);
+    break;
+  case Utility::HistogramBucketsValue::Cumulative:
+    statsAsJsonHistogramBucketsHelper(
+        *histograms_obj_container_fields, all_histograms, used_only, regex,
+        [](const Stats::ParentHistogramSharedPtr& histogram) {
+          return histogram->intervalStatistics().computedBuckets();
```

Comment:
both `intervalStatistics()` and `cumulativeStatistics()` return type `HistogramStatistics&` so I think you can clean this up by writing the functor as
```
  [](const Stats::HistogramStatistics& hist_stats) -> std::vector<uint64_t> { return stats.computedBuckets(); }
```
and then you only need to pass in one of them.

I'm trying to keep this clean and minimal as I have another refactor coming in this code that this will conflict with. I think yours may land first, but FYI mine is: https://github.com/envoyproxy/envoy/pull/19693 and the smaller and cleaner we can make this, the less trouble it will be when we have to atomize it and reconstruct it for the streaming support :)

Reply:
Done